ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -558,37 +554,43 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr = do
         curTip      = castPoint $ AF.headPoint curChain
         curHead     = AF.headAnchor curChain
 
-    -- | We have found a path of hashes to the new block through the
-    -- VolatileDB. We try to extend this path by looking for forks that start
-    -- with the given block, then we do chain selection and /possibly/ try to
-    -- switch to a new fork.
+    -- | We have found a 'PointChainDiff' through the VolatileDB connecting
+    -- the new block to the current chain. We'll call the intersection/anchor
+    -- @x@.
+    --
+    -- We try to extend this path by looking for forks that start with the
+    -- given block, then we do chain selection and /possibly/ try to switch to
+    -- a new fork.
     switchToAFork :: HasCallStack
                   => (WithOrigin (HeaderHash blk) -> Set (HeaderHash blk))
                   -> ChainAndLedger blk
                      -- ^ The current chain (anchored at @i@) and ledger
-                  -> NonEmpty (HeaderHash blk)
-                     -- ^ An uninterrupted path of hashes @(i,b]@.
+                  -> PointChainDiff blk
+                     -- ^ Path of points @(x,b]@
                   -> m (Point blk)
-    switchToAFork succsOf curChainAndLedger hashes = do
+    switchToAFork succsOf curChainAndLedger pdiff = do
         let suffixesAfterB = VolDB.candidates succsOf (realPointToPoint p)
-            initCache      = Map.insert (headerHash hdr) hdr (cacheHeaders curChain)
-        -- Fragments that are anchored at @i@.
-        candidates <- flip evalStateT initCache $
-          case NE.nonEmpty suffixesAfterB of
-            -- If there are no suffixes after @b@, just use the fragment that
-            -- ends in @b@ as the sole candidate.
-            Nothing              -> (NE.:| []) <$> constructFork i hashes []
-            Just suffixesAfterB' -> mapM (constructFork i hashes . NE.toList)
-                                         suffixesAfterB'
-        let chainDiffs =
-                -- The suffixes all fork off from the current chain within @k@
-                -- blocks, so it satisfies the precondition of
-                -- 'preferCandidate'.
-                filter
-                  (preferAnchoredCandidate cdbTopLevelConfig curChain .
-                   Diff.getSuffix)
-              . mapMaybe (Diff.diff curChain)
-              $ NE.toList candidates
+            -- We use a cache to avoid reading the headers from disk multiple
+            -- times in case they're part of multiple forks that go through
+            -- @b@.
+            initCache      =  Map.singleton (headerHash hdr) hdr
+        chainDiffs <-
```

Comment:
The five steps:

1. Compute candidates (`VolDB.candidates`) as `AnchoredFragment`s
2. Extend our diffs with them
3. Filter on suffix >= rollback
4. Translate diffs from `HeaderFields` to headers
5. Filter on `preferAnchoredCandidate`